Resources/ProtobufDumper/ProtobufDumper/ProtobufDumper.cs

```diff
@@ -0,0 +1,957 @@
+                var protoNode = new ProtoNode()
+                {
+                    Name = proto.name,
+                    Proto = proto,
+                    Dependencies = new List<ProtoNode>(),
+                    AllPublicDependencies = new HashSet<FileDescriptorProto>(),
+                    Types = new List<ProtoTypeNode>(),
+                    Defined = true
+                };
+
+                protobufMap.Add( proto.name, protoNode );
+
+                foreach ( var extension in proto.extension )
+                {
+                    protoNode.Types.Add( GetOrCreateTypeNode( GetPackagePath( proto.package, extension.name ), proto, extension ) );
+
+                    if ( IsNamedType( extension.type ) && !string.IsNullOrEmpty( extension.type_name ) )
+                        protoNode.Types.Add( GetOrCreateTypeNode( GetPackagePath( proto.package, extension.type_name ) ) );
+
+                    if ( !string.IsNullOrEmpty( extension.extendee ) )
+                        protoNode.Types.Add( GetOrCreateTypeNode( GetPackagePath( proto.package, extension.extendee ) ) );
+                }
+
+                foreach ( var enumType in proto.enum_type )
+                {
+                    protoNode.Types.Add( GetOrCreateTypeNode( GetPackagePath( proto.package, enumType.name ), proto, enumType ) );
+                }
+
+                foreach ( var messageType in proto.message_type )
+                {
+                    RecursiveAnalyzeMessageDescriptor( messageType, protoNode, proto.package );
+                }
+
+                foreach ( var service in proto.service )
+                {
+                    protoNode.Types.Add( GetOrCreateTypeNode( GetPackagePath( proto.package, service.name ), proto, service ) );
+
+                    foreach ( var method in service.method )
+                    {
+                        if ( !string.IsNullOrEmpty( method.input_type ) )
+                            protoNode.Types.Add( GetOrCreateTypeNode( GetPackagePath( proto.package, method.input_type ) ) );
+
+                        if ( !string.IsNullOrEmpty( method.output_type ) )
+                            protoNode.Types.Add( GetOrCreateTypeNode( GetPackagePath( proto.package, method.output_type ) ) );
+                    }
+                }
+            }
+
+            // inspect file dependencies
+            var missingDependencies = new List<ProtoNode>();
+
+            foreach ( var pair in protobufMap )
+            {
+                foreach ( var dependency in pair.Value.Proto.dependency )
+                {
+                    if ( dependency.StartsWith( "google", StringComparison.Ordinal ) ) continue;
+
+                    if ( protobufMap.TryGetValue( dependency, out var depends ) )
+                    {
+                        pair.Value.Dependencies.Add( depends );
+                    }
+                    else
+                    {
+                        Console.ForegroundColor = ConsoleColor.Cyan;
```

Comment:
More UI mixed in with analysis logic

Reply:
Not addressed yet. This needs to inspect some of the gory internals